### PR TITLE
v3 to v4: injectGlobalStyle to createGlobalStyle

### PR DESCRIPTION
**Pull request by StevenLangbroek**
## Todo

- [x] Prepare file structure for additional codemods
- [x] Bootstrap `injectGlobalStyle` to `createGlobalStyle`
- [x] Implement
- [x] ???
- [x] Profit

Fixes styled-components/styled-components#1961

### CODE REVIEW

**Comment by mxstbr** [comment]
Excited! Can't wait to get this in

**Comment by probablyup** [comment]
Let me know if you need any help here Steven. I'm back home now 👍

**Comment by StevenLangbroek** [comment]
Thanks @probablyup! (and welcome back :) ) I have 2 open questions:

- Can you help me write some test cases? I covered the basic one now.
- I can imagine doing some cleanup or just shipping now, either way is fine by me. If you want to see some polish, let me know and I'll clean it up.

File: src/v4/injectGlobalToCreateGlobalStyle.js [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 
 const t = require("babel-core").types;
-const { prop } = require("lodash/fp");
+const { prop, set } = require("lodash/fp");
 const { manipulateOptions } = require("./common");
 
 const WARNING =
   "An injectGlobal usage was converted to createGlobalStyles via codemod but needs to be hooked up. See styled-components.com/docs/api#createglobalstyle for instructions.";
-
 const PREVIOUS = "injectGlobalStyle";
```
Comment (probablyup) [inline src/v4/injectGlobalToCreateGlobalStyle.js]: It's just "injectGlobal", not "injectGlobalStyle"

**Review by quantizor — COMMENTED** [review]
Tests:

1. import injectGlobal via ES import
```js
import { injectGlobal } from 'styled-components'

injectGlobal`
  html { color: red; }
`
```
1. import injectGlobal via CJS require and use via property
```js
const styled = require('styled-components');
styled.injectGlobal`
  html { color: red; }
`
```
1. import injectGlobal via CJS require w/ destructuring
```js
const { injectGlobal } = require('styled-components')

injectGlobal`
  html { color: red; }
`
```

I'd just do snapshot tests for each to verify they're rewritten correctly

File: src/v4/injectGlobalToCreateGlobalStyle.js (hunk 1 above)
Comment (StevenLangbroek) [reply src/v4/injectGlobalToCreateGlobalStyle.js]: Haha yep, thanks.

**Comment by StevenLangbroek** [comment]
Alright, will fix later. Thanks for the input!

**Comment by StevenLangbroek** [comment]
Just accounting for the require call and then I'm done. Not super thrilled with the code of the end-result but it works 🤷‍♂️

**Comment by probablyup** [comment]
@StevenLangbroek Yeah I mean I wouldn't worry about it too much - once it's done I doubt we'll ever touch it again

**Comment by StevenLangbroek** [comment]
Boom!

**Comment by StevenLangbroek** [comment]
I could fiddle with this further, but like you said ¯\_(ツ)_/¯

If you think of any edge cases, let me know and I'll account for them.

File: src/v4/injectGlobalToCreateGlobalStyle.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
+"use strict";
+
+const t = require("babel-core").types;
+const { prop, set, has, flow, findIndex } = require("lodash/fp");
+const { manipulateOptions } = require("./common");
+
+const WARNING =
+  "An injectGlobal usage was converted to createGlobalStyles via codemod but needs to be hooked up. See styled-components.com/docs/api#createglobalstyle for instructions.";
```
Comment (probablyup) [inline src/v4/injectGlobalToCreateGlobalStyle.js]: Can you make this a full URL with https etc?

**Review by quantizor — APPROVED** [review]
Looking fresh AF. Thanks for all the work!

File: src/v4/injectGlobalToCreateGlobalStyle.js (hunk 2 above)
Comment (StevenLangbroek) [reply src/v4/injectGlobalToCreateGlobalStyle.js]: Done.

**Comment by StevenLangbroek** [comment]
@mxstbr what about the infra? Do I need to update `manualRun.js` for example?

**Comment by probablyup** [comment]
Gotta update the snaps after the text change

**Comment by StevenLangbroek** [comment]
Doh!

**Comment by probablyup** [comment]
Seems like `manualRun` could be replaced with tests?

**Comment by StevenLangbroek** [comment]
I'm not sure it's needed, assuming babel-codemod provides some dry-run capabilities?

**Comment by probablyup** [comment]
Yup it does have one. We can probably ditch that file. Could you add your transform to the README?

**Comment by StevenLangbroek** [comment]
On it!

**Comment by StevenLangbroek** [comment]
Hmm... any thoughts how to structure this for commander? I don't have much experience with it. Do we let people run 2 commands (v4-extend, v4-global)?

**Comment by StevenLangbroek** [comment]
This is what I have now, it's kinda elegant:

```js
program
  .version(pkg.version)
  .command("v4-extend [files...]")
  .action(runCodemod('v4', 'extendToStyled'))
  .command("v4-global [files...]")
  .action(runCodemod('v4', 'injectGlobalToCreateGlobalStyle'))
```

**Comment by probablyup** [comment]
Ideally we'd have a simple "v4" target that runs all v4 mods, and then individual commands for each mod if someone wants to run them individually.

**Comment by StevenLangbroek** [comment]
I could turn it into something like:

```js
registerCodemods("v4", ["extendToStyled", "injectGlobalToCreateGlobalStyle"])(program)
```

**Comment by StevenLangbroek** [comment]
Oh boy. Ok this is a bit hairier than I thought: if I want to run all the codemods I have to run them sequentially, which... is... harsh.

**Comment by StevenLangbroek** [comment]
Have any better ideas than throwing `async` module at it?

File: bin/styled-components-codemods.js [hunk 3]
```diff
@@ -1,44 +1,75 @@
...
+const program = require("commander");
+const { spawnSync } = require("child_process");
+const { curry } = require("lodash");
+const pkg = require("../package.json");
+
+/**
+ * Add the version-specific codemods here.
+ */
+const codemods = {
+  v4: ["extendToStyled", "injectGlobalToCreateGlobalStyle"]
```
Comment (StevenLangbroek) [inline bin/styled-components-codemods.js]: The upshot of this is all you have to do to wire up another one to commander is add its filename here.

Comment (StevenLangbroek) [reply bin/styled-components-codemods.js]: I'm thinking later we can standardize on this more (e.g., examples are still disconnected from this structure).

**Comment by StevenLangbroek** [comment]
@probablyup @mxstbr I'm not sure what's left here...

**Comment by probablyup** [comment]
It seems to be adding an error message to every file it runs over rather than just the one that uses injectGlobal. You can see if you run it over one of my test repos: https://github.com/probablyup/cra-styled-components-font-example

**Review by mxstbr — APPROVED** [review]
Don't have time for a thorough review rn, but if it works and @probablyup approves let's ship it!

**Comment by quantizor** [comment]
Thanks @StevenLangbroek!

**Pull request merged**